### PR TITLE
Remove DAM annotations from enum converter

### PR DESCRIPTION
**Pull request by LakshanF**
Remove DAM annotations from Enum converter since the trimmer always preserve `enums`.

### CODE REVIEW

**Review by sbomer — APPROVED** [review]
LGTM but I'd appreciate if @MichalStrehovsky could take a look too.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs [hunk 1]
```diff
@@ -20,12 +21,12 @@ public class EnumConverter : TypeConverter
         /// Initializes a new instance of the <see cref='System.ComponentModel.EnumConverter'/> class for the given
         /// type.
         /// </summary>
-        public EnumConverter([DynamicallyAccessedMembers(TypeDescriptor.ReflectTypesDynamicallyAccessedMembers)] Type type)
+        public EnumConverter(Type type)
         {
+            Debug.Assert(type.IsEnum || type.Equals(typeof(Enum)), "type should be an Enum type");
```
Comment (MichalStrehovsky) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:26]: This is public API - need argument validation not assertion. The code in the instance method seems to assume .IsEnum

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs [hunk 2]
```diff
@@ -115,6 +116,7 @@ private static long GetEnumValue(bool isUnderlyingTypeUInt64, object enumVal, Cu
         /// <summary>
         /// Converts the given value object to the specified destination type.
         /// </summary>
+        [UnconditionalSuppressMessage("Trimming", "IL2075:", Justification = "Trimmer does not trim enums")]
```
Comment (MichalStrehovsky) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:119]: These suppressions are on very long methods. We prefer as targeted suppressions as possible. E.g. `[UnconditionalSuppress] FieldInfo GetPublicStaticEnumField(string name) => EnumType.GetField(name, Public | Static)`.

**Review by MichalStrehovsky — CHANGES_REQUESTED** [review]

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs (hunk 2 above)
Comment (LakshanF) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs]: Thanks!

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs (hunk 1 above)
Comment (LakshanF) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:26]: Now throwing

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs [hunk 3]
```diff
@@ -227,9 +236,15 @@ public override StandardValuesCollection GetStandardValues(ITypeDescriptorContex
                 // We need to get the enum values in this rather round-about way so we can filter
                 // out fields marked Browsable(false). Note that if multiple fields have the same value,
                 // the behavior is undefined, since what we return are just enum values, not names.
-                Type reflectType = TypeDescriptor.GetReflectionType(EnumType) ?? EnumType;
+                [UnconditionalSuppressMessage("Trimming", "IL2067:", Justification = "Trimmer does not trim enums")]
+                static Type GetTypeDescriptorReflectionType(Type enumType) => TypeDescriptor.GetReflectionType(enumType);
+
+                Type reflectType = GetTypeDescriptorReflectionType(EnumType) ?? EnumType;
```
Comment (MichalStrehovsky) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:242]: Ok, this is the kind of thing that would be hidden if we just left this as suppressed on the whole method. Does anything guarantee `TypeDescriptor.GetReflectionType` returns an enum? If I'm reading the code correctly, this is looking at fields of whatever `GetReflectionType` returned and trying to find fields with the same name on the original enum. Looking at the enum is fine, but I don't know how to reason about `GetReflectionType`.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs [hunk 4]
```diff
@@ -227,9 +236,15 @@ public override StandardValuesCollection GetStandardValues(ITypeDescriptorContex
...
+
+                [UnconditionalSuppressMessage("Trimming", "IL2070:", Justification = "Trimmer does not trim enums")]
+                static FieldInfo[]? GetPublicStaticEnumFields(Type type) => type.GetFields(BindingFlags.Public | BindingFlags.Static);
```
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:245]: If we are going to propagate this suppression throughout the stack, would it make sense to make new APIs specifically for Enums? There are other places that get the FieldInfo of enums. For example: 

https://github.com/microsoft/kiota-abstractions-dotnet/blob/9852c37689c53c33ba91dbeb8a74dfe1e555d4f8/src/RequestInformation.cs#L209-L214

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs (hunk 3 above)
Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:242]: This code and comment is a little odd. If we think we already have an Enum type, why do we need to call into TypeDesciptor.GetReflectionType?

It might make sense to go back into the source history for .NET Framework and dig at exactly what scenario this code is enabling.

My concern is that someone is passing in a `Type` into `new EnumConverter` that isn't an Enum, but instead has a `TypeDescriptionProviderAttribute` on it that points to an Enum type. We may be breaking a scenario here with this change.

Comment (MichalStrehovsky) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:242]: Yep, we feed EnumType (the thing passed to constructor) to a lot of Enum APIs (not just parse, but also Enum.ToObject, Enum.GetUnderlyingType, Enum.IsDefined) so this is not going to work well for non-enum.

However, the type returned from `TypeDescriptor.GetReflectionType` doesn't seem to be fed to enum APIs and only used in this one spot. I don't know how to make sense of it.

Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:242]: What I'm ultimately asking for here is to figure out why this code even exists. We already have the enum Type (we think), why are we calling to `TypeDescriptor.GetReflectionType`? Can the `Type` in the constructor not be a RuntimeType? And calling `TypeDescriptor.GetReflectionType` gets the RuntimeType?

Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:242]: > I didn't see any explanation in the https://github.com/dotnet/corefx/pull/7454 that added this either.

That change just copied the code from .NET Framework. You can see the exact same comment here: https://referencesource.microsoft.com/#System/compmod/system/componentmodel/EnumConverter.cs,247

Comment (MichalStrehovsky) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:242]: > `GetReflectionType` is already annotated as requiring and returning PublicParameterlessConstructor | PublicFields so I think we've got that covered.

Didn't realize this part - if that's the case, can we make it so that the result of GetReflectionType doesn't go through the suppressed codepath? I.e. if someone in the future decides to drop the `return:` annotation on this API, we'd get a warning here.

Basically right now `GetPublicStaticEnumFields` assumes the suppression is safe because it's an enum. But the return value from `GetReflectionType` doesn't need a suppression in the first place.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs (hunk 4 above)
Comment (LakshanF) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs]: Created #100592 to track this.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs [hunk 5]
```diff
@@ -227,9 +235,27 @@ public override StandardValuesCollection GetStandardValues(ITypeDescriptorContex
                 // We need to get the enum values in this rather round-about way so we can filter
                 // out fields marked Browsable(false). Note that if multiple fields have the same value,
                 // the behavior is undefined, since what we return are just enum values, not names.
-                Type reflectType = TypeDescriptor.GetReflectionType(EnumType) ?? EnumType;
+                // Given that EnumType is constrained to be an enum, we suppress calls for reflection with Enum.
+
+                [UnconditionalSuppressMessage("Trimming", "IL2067:", Justification = "Trimmer does not trim Enums")]
+                [return: DynamicallyAccessedMembers(TypeDescriptor.ReflectTypesDynamicallyAccessedMembers)]
```
Comment (steveharter) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:241]: Do you know why [TypeDescriptor.ReflectTypesDynamicallyAccessedMembers](https://github.com/dotnet/runtime/blob/e629779764abd70bde54e780eeeb8fd5bee84fac/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs#L22) is defined as
```DynamicallyAccessedMemberTypes.PublicParameterlessConstructor | DynamicallyAccessedMemberTypes.PublicFields;```
instead of
```DynamicallyAccessedMemberTypes.PublicParameterlessConstructor | DynamicallyAccessedMemberTypes.PublicProperties;``` ?

cc @eerhardt

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs [hunk 6]
```diff
@@ -20,12 +21,16 @@ public class EnumConverter : TypeConverter
         /// Initializes a new instance of the <see cref='System.ComponentModel.EnumConverter'/> class for the given
         /// type.
         /// </summary>
-        public EnumConverter([DynamicallyAccessedMembers(TypeDescriptor.ReflectTypesDynamicallyAccessedMembers)] Type type)
+        public EnumConverter(Type type)
         {
+            if (!type.IsEnum && !type.Equals(typeof(Enum)))
+            {
+                throw new ArgumentException(SR.EnumInvalidValue);
```
Comment (steveharter) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:28]: Is this an error case you found?

Comment (steveharter) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:28]: When would someone pass in `typeof(Enum)`?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs (hunk 5 above)
Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:241]: I believe it is because these were the only things that were needed to be preserved because this was only called in 2 places:

https://github.com/dotnet/runtime/blob/470df6172572054bdcdd4f798f9a34abd282dd49/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs#L260-L269

https://github.com/dotnet/runtime/blob/470df6172572054bdcdd4f798f9a34abd282dd49/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs#L230-L232

Since these places only needed ctor and fields, that is all that was guaranteed to be preserved.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs (hunk 6 above)
Comment (LakshanF) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:28]: Yep, the abstract Enum class seems to be supported in `EnumConverter` and there are [tests ](https://github.com/dotnet/runtime/blob/main/src/libraries/System.ComponentModel.TypeConverter/tests/EnumConverterTest.cs#L35)for this.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs [hunk 7]
```diff
@@ -20,12 +21,16 @@ public class EnumConverter : TypeConverter
         /// Initializes a new instance of the <see cref='System.ComponentModel.EnumConverter'/> class for the given
         /// type.
         /// </summary>
-        public EnumConverter([DynamicallyAccessedMembers(TypeDescriptor.ReflectTypesDynamicallyAccessedMembers)] Type type)
+        public EnumConverter(Type type)
         {
+            if (!type.IsEnum && !type.Equals(typeof(Enum)))
```
Comment (MichalStrehovsky) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:26]: I'm a little bit concerned we have tests where this is used with a non-enum and it apparently works (System.Enum is really no different from just System.Object, or whatever).

**Review by MichalStrehovsky — APPROVED** [review]
Looks good from trimming perspective. Can't talk about TypeConverter or why we accept System.Enum (and whether this should accept all non-enums for the same reasons it accepts System.Enum).

**Comment by ericstj** [comment]
Seems there are a few others who might be broken by this: https://github.com/search?q=%2Fbase%5C(type%5C.GetType%5C(%5C)%5C)%2F&type=code
